### PR TITLE
Return TIMESTAMP WITH TIME ZONE from Prometheus connector 

### PR DESCRIPTION
**Pull request by dain**

### CODE REVIEW

File: presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusClient.java [hunk 1]
```diff
@@ -52,7 +52,7 @@
 
 public class PrometheusClient
 {
-    static final Type TIMESTAMP_COLUMN_TYPE = TimestampType.createTimestampType(3);
+    static final Type TIMESTAMP_COLUMN_TYPE = TimestampWithTimeZoneType.createTimestampWithTimeZoneType(3);
```
Comment (electrum) [inline presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusClient.java:55]: Static import

File: presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusClock.java [hunk 2]
```diff
@@ -29,21 +29,21 @@
...
     {
-        return LocalDateTime.now(clock);
+        return clock.instant();
     }
 
-    public static PrometheusClock fixedClockAt(LocalDateTime date)
+    public static PrometheusClock fixedClockAt(Instant fixedInstant)
     {
-        return new PrometheusClock(Clock.fixed(date.atZone(ZoneId.systemDefault()).toInstant(), ZoneId.systemDefault()));
+        return new PrometheusClock(Clock.fixed(fixedInstant, ZoneId.of("UTC")));
```
Comment (electrum) [inline presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusClock.java:47]: You can use `ZoneOffset.UTC` which is a `ZoneId`

File: presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusRecordCursor.java [hunk 3]
```diff
@@ -139,8 +141,10 @@ public long getLong(int field)
     {
         Type type = getType(field);
         if (type.equals(TIMESTAMP_COLUMN_TYPE)) {
-            Timestamp timestamp = (Timestamp) requireNonNull(getFieldValue(field));
-            return timestamp.toInstant().toEpochMilli();
+            Instant dateTime = (Instant) requireNonNull(getFieldValue(field));
+            // render with the fixed offset of the Presto server
```
Comment (electrum) [inline presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusRecordCursor.java:145]: This seems strange the data is coming from a different system (unlikely JMX where it's local to that JVM). Using session time zone seems better. Though in practice these are likely to be the same.

File: presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusClient.java [hunk 4]
```diff
@@ -13,75 +13,82 @@
...
+        tableSupplier = Suppliers.memoizeWithExpiration(
+                () -> fetchMetrics(metricCodec, prometheusMetricsUri),
+                config.getCacheDuration().toMillis(),
+                MILLISECONDS);
+        varcharMapType = typeManager.getType(mapType(VARCHAR.getTypeSignature(), VARCHAR.getTypeSignature()));
     }
 
-    private URI getPrometheusMetricsURI(PrometheusConnectorConfig config)
-            throws URISyntaxException
+    private static URI getPrometheusMetricsURI(URI prometheusURI)
```
Comment (electrum) [inline presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusClient.java:79]: Lowercase `Uri`

File: presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusClient.java [hunk 5]
```diff
@@ -13,75 +13,82 @@
...
-        // endpoint to retrieve metric names from Prometheus
-        URI uri = config.getPrometheusURI();
-        return new URIBuilder()
-                .setScheme(uri.getScheme())
-                .setHost(uri.getAuthority())
-                .setPath(uri.getPath().concat(METRICS_ENDPOINT))
-                .build();
+        try {
+            // endpoint to retrieve metric names from Prometheus
+            return new URIBuilder()
```
Comment (electrum) [inline presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusClient.java:83]: We should get rid of this usage of `org.apache.http.client.utils.URIBuilder`. There's another PR #4698 to use Airlift HTTP client.
```java
return new URI(uri.getScheme(), uri.getAuthority(), uri.getPath() + METRICS_ENDPOINT, null, null);

File: presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusClient.java [hunk 6]
```diff
@@ -106,72 +113,59 @@ public PrometheusTable getTable(String schema, String tableName)
...
-            getBearerAuthInfoFromFile().map(bearerToken ->
-                    requestBuilder.header("Authorization", "Bearer " + bearerToken));
-        }
-        requestBuilder.url(uri.toURL());
-        Request request = requestBuilder.build();
-        Response response = httpClient.newCall(request).execute();
-        if (response.isSuccessful()) {
-            return response.body();
+        Request.Builder requestBuilder = new Request.Builder()
+                .url(uri.toString());
```
Comment (electrum) [inline presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusClient.java:143]: No need to wrap here

File: presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusClient.java [hunk 7]
```diff
@@ -106,72 +113,59 @@ public PrometheusTable getTable(String schema, String tableName)
...
-                    }
-                    catch (Exception e) {
-                        throw new PrestoException(NOT_FOUND, "Failed to find/read file: " + tokenFileName, e);
-                    }
-                }).orElse(Optional.empty());
+        return bearerTokenFile.map(tokenFileName -> {
+            try {
+                return readString(tokenFileName.toPath(), UTF_8);
+            }
+            catch (Exception e) {
```
Comment (electrum) [inline presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusClient.java:166]: `IOException`

File: presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusClient.java [hunk 8]
```diff
@@ -106,72 +113,59 @@ public PrometheusTable getTable(String schema, String tableName)
...
+                throw new PrestoException(NOT_FOUND, "Failed to find/read file: " + tokenFileName, e);
```
Comment (electrum) [inline presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusClient.java:167]: Let's make this
```java
throw new PrestoException(PROMETHEUS_UNKNOWN_ERROR, "Failed to read bearer token file: " + tokenFileName, e);

File: presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusMetadata.java [hunk 9]
```diff
@@ -171,9 +171,8 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
     @Override
     public Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(ConnectorSession session, ConnectorTableHandle handle, Constraint constraint)
     {
-        PrometheusTableHandle promHandle = (PrometheusTableHandle) handle;
-        promHandle = new PrometheusTableHandle(promHandle.getSchemaName(), promHandle.getTableName());
-        promHandle.setPredicate(Optional.of(constraint.getSummary()));
+        PrometheusTableHandle promHandle = ((PrometheusTableHandle) handle)
```
Comment (electrum) [inline presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusMetadata.java:174]: Fix abbreviation

File: presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusQueryResponseParse.java [hunk 10]
```diff
@@ -105,14 +95,13 @@ private boolean parse()
                     break;
                 case "scalar":
                 case "string":
-                    stringOrScalarResult = mapper.readValue(result, new TypeReference<PrometheusTimeSeriesValue>() {});
+                    PrometheusTimeSeriesValue stringOrScalarResult = mapper.readValue(result, new TypeReference<PrometheusTimeSeriesValue>() {});
                     Map<String, String> madeUpMetricHeader = new HashMap<>();
                     madeUpMetricHeader.put("__name__", resultType);
-                    PrometheusTimeSeriesValueArray timeSeriesValues = new PrometheusTimeSeriesValueArray(Arrays.asList(stringOrScalarResult));
-                    results = Arrays.asList(new PrometheusMetricResult(madeUpMetricHeader, timeSeriesValues));
+                    PrometheusTimeSeriesValueArray timeSeriesValues = new PrometheusTimeSeriesValueArray(Collections.singletonList(stringOrScalarResult));
```
Comment (electrum) [inline presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusQueryResponseParse.java:101]: Static import

File: presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusRecordCursor.java (hunk 3 above)
Comment (dain) [reply presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusRecordCursor.java:145]: I think it should either be JVM or a configuration.  We can add a configuration later if people want.

**Pull request merged**